yateto/codegen/copyscaleadd/csa_gen.py

```diff
@@ -0,0 +1,176 @@
+from ..common import *
+from ..cache import RoutineGenerator, GpuRoutineGenerator
+from gemmforge import DenseMatrix, CsaGenerator, GenerationError
+from gemmforge import arch as GemmForgeArch
+import re
+
+from yateto.type import Tensor
+
+
+class CsaGen(object):
+  def __init__(self, arch, descr):
+    self._arch = arch
+    self._descr = descr
+
+  def _formatTerm(self, alpha, term):
+    """Generate a sub-string of a term for a source code which is going to be used
+    inside of the inner most for-loop
+
+    Args:
+      alpha (Union[Scalar, float]): TODO
+      term (IndexedTensorDescription): TODO
+
+    Returns:
+
+    Examples:
+      >>> from yateto.memory import DenseMemoryLayout
+      >>> from yateto.ast.indices import Indices
+      >>> from yateto.codegen.common import IndexedTensorDescription
+      >>> from yateto.aspp import dense
+      >>> from yateto.codegen.copyscaleadd.generic import Generic
+      >>> tensor_shape = (5, 6)
+      >>> layout = DenseMemoryLayout(shape=tensor_shape)
+      >>> indices = Indices(indexNames='ij', shape=tensor_shape)
+      >>> description = IndexedTensorDescription(name='A', \
+                                                 indices=indices, \
+                                                 memoryLayout=layout, \
+                                                 eqspp=dense(shape=tensor_shape))
+      >>> obj = Generic(arch='dummy', descr=description)
+      >>> obj._formatTerm(alpha=3, term=description)
+      '3 * A[1*i + 5*j]'
+    """
+
+    prefix = ''
+    if alpha == 0.0:
+      return ''
+
+    if alpha == 1.0:
+      prefix = term.name
+    else:
+      prefix = '{} * {}'.format(alpha, term.name)
+
+    return '{}[{}]'.format(prefix, term.memoryLayout.addressString(term.indices))
+
+  def generate(self, cpp, routineCache):
+    """Generates a tensor equation of a form: B = beta * B + alpha * A
+    Args:
+      cpp (IO): a file stream
+      routineCache:
+
+    Returns:
+
+    """
+
+    d = self._descr  # type: copyscaleadd.Description
+    m = d.loopRanges[d.result.indices[0]]
+    n = d.loopRanges[d.result.indices[1]]
+
+    # TODO:  implement initialization in the generator
+    #if d.beta == 0.0:
+    #  writeBB = boundingBoxFromLoopRanges(d.result.indices, d.loopRanges)
+    #  initializeWithZero(cpp, self._arch, d.result, writeBB)
+
+    # calculate FLOP per single operation
+    flop = 0
+    alpha = d.alpha
+
+    if alpha not in [-1.0, 1.0]:
+      flop += 1
+
+    if d.beta == 1.0 and alpha == -1.0:
+      alpha = 1.0
+      flop += 1
+
+    elif d.beta == 1.0:
+      flop += 1
+
+    elif d.beta != 0.0:
+      raise NotImplementedError
```

Comment:
From python reference manual:

```
raise_stmt:     "raise" [expression ["," expression ["," expression]]]
```


*If no expressions are present, raise re-raises the last expression that was raised in the current scope.*

*Otherwise, raise evaluates its first expression, which must yield a string, class, or instance object ... If the first object is a class or instance, it raises the exception identified by the class of the instance determined in the previous step, with the instance as its parameter.*

In short, you can raise either a class or an object. Anyway, a trace in case of an error will show you where a processes stops and it is more then enough for debugging